lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -0,0 +1,224 @@
+//===- SVExtractTestCode.cpp - SV Simulation Extraction Pass --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass extracts simulation constructs to submodules.  It
+// will take simulation operations, write, finish, assert, assume, and cover and
+// extract them and the dataflow into them into a separate module.  This module
+// is then instantiated in the original module.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Dialect/SV/SVVisitors.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+
+#include <set>
+
+using namespace mlir;
+using namespace circt;
+using namespace sv;
+
+//===----------------------------------------------------------------------===//
+// StubExternalModules Helpers
+//===----------------------------------------------------------------------===//
+
+// Compute the dataflow for a set of ops.
+static void dataflowSlice(ArrayRef<Operation *> ops,
+                          SetVector<Operation *> &results) {
+  for (auto op : ops) {
+    getBackwardSlice(op, &results, [](Operation *testOp) -> bool {
```

Comment:
I'd recommend writing this iterative instead of recursive.  Unbounded recursion always eventually explodes the stack at the worst time.